cpp/src/lists/contains.cu

```diff
@@ -44,296 +41,241 @@
+
+    auto out_positions = make_numeric_column(
+      data_type{type_to_id<size_type>()}, lists.size(), cudf::mask_state::UNALLOCATED, stream, mr);
+    auto out_validity   = rmm::device_uvector<bool>(lists.size(), stream);
+    auto const out_iter = thrust::make_zip_iterator(
+      out_positions->mutable_view().template begin<size_type>(), out_validity.begin());
+
+    auto const keys_iter = cudf::detail::make_optional_iterator<Type>(
+      *keys_dv_ptr, nullate::DYNAMIC{search_keys_have_nulls});
+
+    thrust::tabulate(rmm::exec_policy(stream),
+                     out_iter,
+                     out_iter + lists.size(),
+                     search_lists_fn<Type, decltype(keys_iter)>{lists_cdv, keys_iter, find_option});
+
+    if (search_keys_have_nulls || lists.has_nulls()) {
+      auto [null_mask, null_count] = cudf::detail::valid_if(
+        out_validity.begin(), out_validity.end(), thrust::identity{}, stream, mr);
+      out_positions->set_null_mask(std::move(null_mask), null_count);
+    }
+    return out_positions;
   }
 };
 
 /**
- * @brief Converts key-positions vector (from index_of()) to a BOOL8 vector, indicating if
- * the search key was found.
+ * @brief Converts key-positions vector (from `index_of()`) to a BOOL8 vector, indicating if
+ *        the search key(s) were found.
  */
 std::unique_ptr<column> to_contains(std::unique_ptr<column>&& key_positions,
                                     rmm::cuda_stream_view stream,
                                     rmm::mr::device_memory_resource* mr)
 {
   CUDF_EXPECTS(key_positions->type().id() == type_id::INT32,
                "Expected input column of type INT32.");
-  // If position == -1, the list did not contain the search key.
-  auto const num_rows        = key_positions->size();
-  auto const positions_begin = key_positions->view().begin<size_type>();
-  auto result =
-    make_numeric_column(data_type{type_id::BOOL8}, num_rows, mask_state::UNALLOCATED, stream, mr);
+  auto const positions_begin = key_positions->view().template begin<size_type>();
+  auto result                = make_numeric_column(
+    data_type{type_id::BOOL8}, key_positions->size(), mask_state::UNALLOCATED, stream, mr);
   thrust::transform(rmm::exec_policy(stream),
                     positions_begin,
-                    positions_begin + num_rows,
-                    result->mutable_view().begin<bool>(),
-                    [] __device__(auto i) { return i != absent_index; });
+                    positions_begin + key_positions->size(),
+                    result->mutable_view().template begin<bool>(),
+                    [] __device__(auto const i) {
+                      // position == NOT_FOUND_SENTINEL: the list does not contain the search key.
+                      return i != NOT_FOUND_SENTINEL;
+                    });
+
+  auto const null_count                    = key_positions->null_count();
   [[maybe_unused]] auto [_, null_mask, __] = key_positions->release();
```

Comment:
It would be good to avoid double underscore in any names.
https://en.cppreference.com/w/cpp/language/identifiers
> the identifiers with a double underscore anywhere are reserved;